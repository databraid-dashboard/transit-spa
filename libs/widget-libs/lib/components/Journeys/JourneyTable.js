var _jsxFileName = '../../src/components/Journeys/JourneyTable.jsx';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import BestJourney from './BestJourney';
import NextBestJourney from './NextBestJourney';

import { fetchJourneys } from '../../actions';

export function timeToLeaveConverter(departureTimeInSeconds) {
  var currentTimeInSeconds = Date.now() / 1000;
  var diff = departureTimeInSeconds - currentTimeInSeconds;

  return Math.floor(diff);
}

export var JourneyTable = function (_Component) {
  _inherits(JourneyTable, _Component);

  function JourneyTable() {
    _classCallCheck(this, JourneyTable);

    return _possibleConstructorReturn(this, (JourneyTable.__proto__ || Object.getPrototypeOf(JourneyTable)).apply(this, arguments));
  }

  _createClass(JourneyTable, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _props = this.props,
          destinationId = _props.destinationId,
          origin = _props.origin,
          destinationsById = _props.destinationsById;

      this.props.fetchJourneys(destinationId, origin, destinationsById[destinationId].address);
    }
  }, {
    key: 'render',
    value: function render() {
      var journeys = this.props.journeys;


      if (!journeys) return React.createElement(
        'div',
        {
          __source: {
            fileName: _jsxFileName,
            lineNumber: 28
          },
          __self: this
        },
        'Loading...'
      );

      var bestJourney = journeys[0];
      var nextBestJourney = journeys[1];

      var timeToLeaveBest = timeToLeaveConverter(bestJourney.departureTimeUTC);
      var timeToLeaveNextBest = timeToLeaveConverter(nextBestJourney.departureTimeUTC);

      return React.createElement(
        'div',
        {
          __source: {
            fileName: _jsxFileName,
            lineNumber: 37
          },
          __self: this
        },
        React.createElement(BestJourney, {
          timeToLeaveInSeconds: timeToLeaveBest,
          steps: bestJourney.transitSteps,
          eta: bestJourney.arrivalTimeText,
          conditionStatus: 'on-time',
          __source: {
            fileName: _jsxFileName,
            lineNumber: 38
          },
          __self: this
        }),
        React.createElement(NextBestJourney, {
          timeToLeaveInSeconds: timeToLeaveNextBest,
          steps: nextBestJourney.transitSteps,
          eta: nextBestJourney.arrivalTimeText,
          conditionStatus: 'future undertain -- see journey table',
          __source: {
            fileName: _jsxFileName,
            lineNumber: 44
          },
          __self: this
        })
      );
    }
  }]);

  return JourneyTable;
}(Component);

JourneyTable.propTypes = {
  destinationId: PropTypes.number.isRequired,
  origin: PropTypes.string.isRequired,
  destinationsById: PropTypes.shape({
    1: PropTypes.object
  }).isRequired,
  fetchJourneys: PropTypes.func.isRequired,
  journeys: PropTypes.arrayOf(PropTypes.object)
};

JourneyTable.defaultProps = {
  destinationId: 1,
  origin: '',
  destinationsById: { 1: {} },
  fetchJourneys: function fetchJourneys() {},
  journeys: [{
    departureTimeUTC: Date.now(),
    arrivalTimeText: '00:00am',
    transitSteps: [{ duration: '1 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
      duration: '23 mins',
      instruction: 'Metro rail towards Warm Springs/South Fremont',
      mode: 'TRANSIT'
    }]
  }, {
    departureTimeUTC: Date.now(),
    arrivalTimeText: '00:00pm',
    transitSteps: [{ duration: '4 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
      duration: '56 mins',
      instruction: 'Metro rail towards Warm Springs/South Fremont',
      mode: 'TRANSIT'
    }]
  }]
};
export var mapStateToProps = function mapStateToProps(state, ownProps) {
  var origin = state.configuration.currentLocation.address;
  var destinationsById = state.destinations.byId;
  var destinationId = ownProps.id;
  var journeys = state.journeys.byDestinationId[destinationId];

  return {
    origin: origin,
    journeys: journeys,
    destinationId: destinationId,
    destinationsById: destinationsById
  };
};

export var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    fetchJourneys: fetchJourneys
  }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(JourneyTable);