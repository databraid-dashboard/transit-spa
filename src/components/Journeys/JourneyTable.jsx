import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Loader } from 'semantic-ui-react';

import JourneyRow from './JourneyRow';

import { fetchJourneys } from '../../actions';
import { timeToLeaveConverter } from '../../utils/journeyTableHelpers';
import injectWidgetId from '../../utils/utils';

export class JourneyTable extends Component {
  constructor(props) {
    super(props);

    this.refreshJourneys = this.refreshJourneys.bind(this);
  }

  componentDidMount() {
    const { destinationId, origin, destinationsById } = this.props;
    this.props.fetchJourneys(destinationId, origin, destinationsById[destinationId].address);
  }

  // TODO: TEST refreshJourneys
  refreshJourneys() {
    const { destinationId, origin, destinationsById } = this.props;
    this.props.fetchJourneys(destinationId, origin, destinationsById[destinationId].address);
  }

  render() {
    const { journeys } = this.props;

    if (!journeys || journeys.length === 0) {
      return <Loader active inline="centered" />;
    }

    const bestJourney = journeys[0];
    const nextBestJourney = journeys[1];
    const bestJourneyStatus = bestJourney.alerts ? bestJourney.alerts[0] : 'on-time';
    const nextBestJourneyStatus = nextBestJourney.alerts ? nextBestJourney.alerts[0] : 'on-time';

    const timeToLeaveBest = timeToLeaveConverter(bestJourney.departureTimeUTC);
    const timeToLeaveNextBest = timeToLeaveConverter(nextBestJourney.departureTimeUTC);

    return (
      <div className="journey-table">
        <JourneyRow
          timeToLeaveInSeconds={timeToLeaveBest}
          steps={bestJourney.transitSteps}
          eta={bestJourney.arrivalTimeText}
          conditionStatus={bestJourneyStatus}
          refreshJourneys={this.refreshJourneys}
        />
        <JourneyRow
          timeToLeaveInSeconds={timeToLeaveNextBest}
          steps={nextBestJourney.transitSteps}
          eta={nextBestJourney.arrivalTimeText}
          conditionStatus={nextBestJourneyStatus}
          refreshJourneys={this.refreshJourneys}
        />
      </div>
    );
  }
}

/* eslint-disable react/require-default-props */
JourneyTable.propTypes = {
  destinationId: PropTypes.number.isRequired,
  origin: PropTypes.string.isRequired,
  destinationsById: PropTypes.shape({
    1: PropTypes.object,
  }).isRequired,
  fetchJourneys: PropTypes.func.isRequired,
  journeys: PropTypes.arrayOf(PropTypes.object),
};

export const mapStateToProps = (state, ownProps) => {
  const id = ownProps.widgetId;
  const origin = state.widgets.byId[id].configuration.currentLocation.address;
  const destinationsById = state.widgets.byId[id].destinations.byId;
  const destinationId = ownProps.id;
  const alerts = state.alerts;
  const journeys = state.widgets.byId[id].journeys.byDestinationId[destinationId];

  return {
    origin,
    journeys,
    destinationId,
    destinationsById,
    alerts,
  };
};

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchJourneys,
    },
    dispatch,
  );

export default injectWidgetId(connect(mapStateToProps, mapDispatchToProps)(JourneyTable));
