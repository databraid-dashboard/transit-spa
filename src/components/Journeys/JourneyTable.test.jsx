import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { JourneyTable, timeToLeaveConverter, mapDispatchToProps } from './JourneyTable';

describe('JourneyTable', () => {
  const testData = {
    destinationId: 1,
    origin: '',
    destinationsById: { 1: {} },
    journeys: [
      {
        departureTimeUTC: 1502222222,
        arrivalTimeText: '00:00am',
        transitSteps: [
          { duration: '1 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' },
          {
            duration: '23 mins',
            instruction: 'Metro rail towards Warm Springs/South Fremont',
            mode: 'TRANSIT',
          },
        ],
      },
      {
        departureTimeUTC: 1502222555,
        arrivalTimeText: '00:00pm',
        transitSteps: [
          { duration: '4 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' },
          {
            duration: '56 mins',
            instruction: 'Metro rail towards Warm Springs/South Fremont',
            mode: 'TRANSIT',
          },
        ],
      },
    ],
    timeToLeaveInSeconds: 1502222223,
  };

  it('should render Loading...', () => {
    const component = shallow(
      <JourneyTable
        id={1}
        destinationId={testData.destinationId}
        origin={testData.origin}
        destinationsById={testData.destinationsById}
        journeys={null}
      />,
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchJourneys');
  });

  it('calls fetchJourneys on load', () => {
    const fetchJourneys = jest.fn();
    mount(
      <JourneyTable
        id={1}
        destinationId={testData.destinationId}
        origin={testData.origin}
        destinationsById={testData.destinationsById}
        journeys={testData.journeys}
        fetchJourneys={fetchJourneys}
      />,
    );

    expect(fetchJourneys).toHaveBeenCalled();
  });
});

describe('timeToLeaveConverter', () => {
  it('should return positive integer', () => {
    const futureTimeInSeconds = Date.now() + 1000;
    expect(timeToLeaveConverter(futureTimeInSeconds)).toBeGreaterThanOrEqual(1);
  });
});
