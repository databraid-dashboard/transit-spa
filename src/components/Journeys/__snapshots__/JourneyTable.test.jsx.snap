// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JourneyTable should render Loading... 1`] = `
<Loader
  active={true}
  inline="centered"
/>
`;
